datafusion/src/execution/runtime_env.rs

```diff
@@ -25,47 +25,110 @@ use crate::{
         memory_manager::{MemoryConsumerId, MemoryManager, MemoryManagerConfig},
     },
 };
-
+use std::fmt;
 use std::fmt::{Debug, Formatter};
+
+use crate::datasource::object_store::{ObjectStore, ObjectStoreRegistry};
+use crate::execution::context::{
+    SessionContextRegistry, TaskContext, TaskContextRegistry, BATCH_SIZE,
+    PARQUET_PRUNING, REPARTITION_AGGREGATIONS, REPARTITION_JOINS, REPARTITION_WINDOWS,
+    TARGET_PARTITIONS,
+};
+use crate::prelude::{SessionConfig, SessionContext};
+use datafusion_common::DataFusionError;
+use once_cell::sync::OnceCell;
+use parking_lot::Mutex;
+use std::path::PathBuf;
 use std::sync::Arc;
 
-#[derive(Clone)]
-/// Execution runtime environment. This structure is passed to the
-/// physical plans when they are run.
+/// Global singleton RuntimeEnv
+pub static RUNTIME_ENV: OnceCell<Arc<RuntimeEnv>> = OnceCell::new();
```

Comment:
I think a single static runtime environment makes sense for Ballista but not for DataFusion (which gets used in a variety of usecases that a single runtime might not be applicable for)

Reply:
> I think a single static runtime environment makes sense for Ballista but not for DataFusion (which gets used in a variety of usecases that a single runtime might not be applicable for)

Hi, Alamb

Thanks a lot for taking a look. One reason that I have to use the global static runtime is make the ExecutionPlans session/configuration aware. 

````
 pub trait ExecutionPlan: Debug + Send + Sync {

    ...................
    
    /// Return the Session id associated with the execution plan.
    fn session_id(&self) -> String;

    /// Return the Session configuration associated with the execution plan.
    fn session_config(&self) -> SessionConfig {
        let session_id = self.session_id();
        let runtime = RuntimeEnv::global();
        runtime.lookup_session_config(session_id.as_str())
    }


impl RuntimeEnv {

    ...................
    
    /// Retrieves a copied version of `SessionConfig` by session_id
    pub fn lookup_session_config(&self, session_id: &str) -> SessionConfig {
        if self.is_local() {
            // It is possible that in a local env such as in unit tests there is no
            // SessionContext created, in this case we have to return a default SessionConfig.
            self.lookup_config(session_id)
                .map_or(SessionConfig::new(), |c| c.lock().clone())
        } else if self.is_scheduler() {
            self.lookup_session(session_id)
                .expect("SessionContext doesn't exist")
                .copied_config()
        } else {
            self.config_from_task_context(session_id)
        }
 }

````

I think even we just use DataFusion as a lib, it is still a valid requirement to make the ExecutionPlan configuration aware. We might add more and more configuration options to enable/disable some feature during plan time and execution runtime(System like SparkSQL/Presto they have more than hundreds of configuration options). We need an extendable way to achieve this. With the single static runtime environment,  I could make the ExecutionPlan configuration aware and runtime aware(it knows it runs in DataFusion local as a lib, or in the Ballista executor/scheduler) without pass-though anything. 

Some of the systems might leverage the ThreadLocal facility in the program language to get the current context/configuration to avoid pass-through configurations. But In Rust we can not leverage ThreadLocal, they are so many async parts in the code paths.

Reply:
Yes, I agree the alternative is having to pass through some sort of context (like the `ExecutionContext`) explicitly everywhere, which may make the code harder to read. However the upside of explicitly passing through any shared state is that there are no surprising behaviors

> I think even we just use DataFusion as a lib, it is still a valid requirement to make the ExecutionPlan configuration aware. We might add more and more configuration options to enable/disable some feature during plan time and execution runtime(System like SparkSQL/Presto they have more than hundreds of configuration options).

Yes I agree having `ExecutionPlan` be configuration aware is important. 

The idea up to this point has been to explicitly copy / plumb through this state (via  `Arc<ExecutionContextState>`)
https://github.com/apache/arrow-datafusion/blob/7a2cbd5500c1e9447c7d71599eeccfdd5833cd4e/datafusion%2Fsrc%2Fexecution%2Fcontext.rs#L148

And  the `RuntimeEnv` on the call to `ExecutionPlan::execute`

https://github.com/apache/arrow-datafusion/blob/33b9357139ad918da0f45a92db37f00ffa64b0ba/datafusion/src/physical_plan/mod.rs#L226-L230


What would you think about adding configuration options to RuntimeEnv (or something else more general)?
https://github.com/apache/arrow-datafusion/blob/641338f726549c10c5bafee34537dc1e56cdec04/datafusion/src/execution/runtime_env.rs#L35-L42